youtube_dl/socks.py

```diff
@@ -0,0 +1,248 @@
+
+        return (dsthost, dstport)
+
+    def _setup_socks5(self, address):
+        destaddr, port = address
+
+        try:
+            ipaddr = socket.inet_aton(destaddr)
+        except socket.error:
+            if self._proxy.remote_dns:
+                ipaddr = None
+            else:
+                ipaddr = socket.inet_aton(socket.gethostbyname(destaddr))
+
+        auth_methods = 1
+        if self._proxy.username and self._proxy.password:
+            # two auth methods available
+            auth_methods = 2
+        packet = struct_pack('!BBB', 0x5, auth_methods, 0x00)  # no auth
+        if self._proxy.username and self._proxy.password:
+            packet += struct_pack('!B', 0x02)  # user/pass auth
+
+        self.sendall(packet)
+
+        packet = self.recvall(2)
+        version, method = struct_unpack('!BB', packet)
+
+        # check valid response
+        if version != 0x05:
+            self.close()
+            raise ProxyError(
+                0, 'Invalid response from server. Expected {0:02x} got {1:02x}'.format(5, version))
+
+        # no auth methods
+        if method == 0xFF:
+            self.close()
+            raise Socks5Error(method)
+
+        # user/pass auth
+        if method == 0x01:
+            username = self._proxy.username
+            if hasattr(username, 'encode'):
+                username = username.encode()
+            password = self._proxy.password
+            if hasattr(password, 'encode'):
+                password = password.encode()
+            packet = struct_pack('!BB', 1, len(username)) + username
+            packet += struct_pack('!B', len(password)) + password
+            self.sendall(packet)
+
+            packet = self.recvall(2)
+            version, status = struct_unpack('!BB', packet)
+
+            if version != 0x01:
+                self.close()
+                raise ProxyError(
+                    0, 'Invalid response from server. Expected {0:02x} got {1:02x}'.format(1, version))
+
+            if status != 0x00:
+                self.close()
+                raise Socks5Error(1)
+        elif method == 0x00:  # no auth
+            pass
+
+        packet = struct_pack('!BBB', 5, 1, 0)
+        if ipaddr is None:
+            if hasattr(destaddr, 'encode'):
+                destaddr = destaddr.encode()
+            packet += struct_pack('!BB', 3, len(destaddr)) + destaddr
+        else:
+            packet += struct_pack('!B', 1) + ipaddr
+        packet += struct_pack('!H', port)
+
+        self.sendall(packet)
+
+        packet = self.recvall(4)
+        version, status, _, atype = struct_unpack('!BBBB', packet)
+
+        if version != 0x05:
+            self.close()
+            raise ProxyError(
+                0, 'Invalid response from server. Expected {0:02x} got {1:02x}'.format(5, version))
+
+        if status != 0x00:
+            self.close()
+            raise Socks5Error(status)
+
+        if atype == 0x01:
```

Comment:
Why not use enum-like constants where possible? It's rather hard to grasp code with magic constants without knowing the internals.